metricbeat/module/linux/memory/data.go

```diff
@@ -18,51 +18,56 @@
 package memory
 
 import (
+	"os"
+	"strconv"
+	"strings"
+
 	"github.com/pkg/errors"
 
 	"github.com/elastic/beats/v7/libbeat/common"
 	"github.com/elastic/beats/v7/libbeat/metric/system/resolve"
 	"github.com/elastic/beats/v7/metricbeat/internal/metrics/memory"
-	sysinfo "github.com/elastic/go-sysinfo"
-	sysinfotypes "github.com/elastic/go-sysinfo/types"
 )
 
 // FetchLinuxMemStats gets page_stat and huge pages data for linux
 func FetchLinuxMemStats(baseMap common.MapStr, hostfs resolve.Resolver) error {
-
-	vmstat, err := GetVMStat()
+	vmstat, err := GetVMStat(hostfs)
 	if err != nil {
 		return errors.Wrap(err, "error fetching VMStats")
 	}
 
-	pageStats := common.MapStr{
-		"pgscan_kswapd": common.MapStr{
-			"pages": vmstat.PgscanKswapd,
-		},
-		"pgscan_direct": common.MapStr{
-			"pages": vmstat.PgscanDirect,
-		},
-		"pgfree": common.MapStr{
-			"pages": vmstat.Pgfree,
-		},
-		"pgsteal_kswapd": common.MapStr{
-			"pages": vmstat.PgstealKswapd,
-		},
-		"pgsteal_direct": common.MapStr{
-			"pages": vmstat.PgstealDirect,
-		},
+	pageStats := common.MapStr{}
+	if pages, ok := vmstat["pgscan_kswapd"]; ok {
+		pageStats.Put("pgscan_kswapd.pages", pages)
+	}
+	if pages, ok := vmstat["pgscan_direct"]; ok {
+		pageStats.Put("pgscan_direct.pages", pages)
+	}
+	if pages, ok := vmstat["pgfree"]; ok {
+		pageStats.Put("pgfree.pages", pages)
+	}
+	if pages, ok := vmstat["pgsteal_kswapd"]; ok {
+		pageStats.Put("pgsteal_kswapd.pages", pages)
 	}
+	if pages, ok := vmstat["pgsteal_direct"]; ok {
+		pageStats.Put("pgsteal_direct.pages", pages)
+	}
+
 	// This is similar to the vmeff stat gathered by sar
 	// these ratios calculate thhe efficiency of page reclaim
-	if vmstat.PgscanDirect != 0 {
+	pgscan, _ := vmstat["pgscan_direct"]
+	pgsteal, pgsteal_ok := vmstat["pgsteal_direct"]
+	if pgscan != 0 && pgsteal_ok {
 		pageStats["direct_efficiency"] = common.MapStr{
-			"pct": common.Round(float64(vmstat.PgstealDirect)/float64(vmstat.PgscanDirect), common.DefaultDecimalPlacesCount),
+			"pct": common.Round(float64(pgsteal)/float64(pgscan), common.DefaultDecimalPlacesCount),
 		}
 	}
 
-	if vmstat.PgscanKswapd != 0 {
+	pgscankswap, _ := vmstat["pgscan_kswapd"]
+	pgstealkswap, pgstealswap_ok := vmstat["pgsteal_kswapd"]
+	if pgscankswap != 0 && pgstealswap_ok {
 		pageStats["kswapd_efficiency"] = common.MapStr{
-			"pct": common.Round(float64(vmstat.PgstealKswapd)/float64(vmstat.PgscanKswapd), common.DefaultDecimalPlacesCount),
+			"pct": common.Round(float64(pgstealkswap)/float64(pgscankswap), common.DefaultDecimalPlacesCount),
```

Comment:
This block and the one above look like a copy paste of each other with tests for each resulting value. Should they also be a helper function?

Reply:
+1